aten/src/ATen/native/cuda/DistributionTemplates.h

```diff
@@ -557,45 +541,20 @@ struct GeometricKernel {
 
 template<typename RNG>
 void exponential_kernel(TensorIterator& iter, double lambda_, RNG gen) {
-  // Note that HIP doesn't support std::nextafter in device code.
-  auto nextafter_1_0_float = std::nextafter(1.0f, 0.0f);
-  auto nextafter_1_0_double = std::nextafter(1.0, 0.0);
   AT_DISPATCH_FLOATING_TYPES_AND2(at::ScalarType::Half, at::ScalarType::BFloat16, iter.dtype(), "exponential_cuda", [&] {
     using accscalar_t = at::acc_type<scalar_t, true>;
     auto lambda = static_cast<accscalar_t>(lambda_);
-    if (std::is_same<scalar_t, double>::value) {
-      // define lambda for exponential transformation
-      auto exponential_func = [lambda, nextafter_1_0_double] __device__ (accscalar_t rand) {
-        accscalar_t sample;
-        // curand_uniform has (0,1] bounds. log(1) is 0 and exponential excludes 0.
-        // Hence, squash the 1 to just below 1.
-        if(rand == static_cast<accscalar_t>(1.0)) {
-          sample = ::log(nextafter_1_0_double);
-        } else {
-          sample = ::log(rand);
-        }
-        return static_cast<scalar_t>(static_cast<accscalar_t>(-1.0) / lambda * sample);
```

Comment:
Logic before: -1 / lambda * log(rand)
Logic after: -1 / lambda * log(1 - rand)

Is this intentional? I don't see description of this in PR description?

Reply:
seems it is wrong, (1-rand) can be 1 for curand-generated rand (smallest rand is 2^-32), so exponential will include 0 that we were trying to avoid.

Reply:
@ezyang I just [fixed it to preserve the original logic](https://github.com/pytorch/pytorch/blob/66417c1a1a938440d5c161d3daada1cf53ccf504/aten/src/ATen/native/cuda/DistributionTemplates.h#L549-L554)

Reply:
Hi, I didn't get to thoroughly read the context behind here but may I know what's meant by `exponential excludes 0`? 

Exponential `variate`, `x` , not exclude 0. 
Exponential `density`, `f(x)`,  excludes 0. x-> infinity, f(x) -> 0.

The random generation is the `variate`, `x`, not the `density`, `f(x)`.  

![exponential_variate](https://user-images.githubusercontent.com/93151422/210113238-abaa071e-0b6e-4907-8b2e-98d414ca00b6.png)





https://en.wikipedia.org/wiki/Exponential_distribution